From 14a7350a92baa0980f3a40f66f4fb0f6d0c0260e Mon Sep 17 00:00:00 2001
From: Christoph Paasch <cpaasch@openai.com>
Date: Wed, 6 Aug 2025 12:23:07 -0700
Subject: [PATCH 08/17] [vxlan] Use proper u16 for vxi_port and fixes to
 vxlan_compare

The kernel expects a u16 for vxi_port, so use NLA_PUT_U16.

The vxlan_compare function is checking on vxi_proxy for too many
fields...
Fix that!

Signed-off-by: Christoph Paasch <cpaasch@openai.com>
Upstream-Status: Backport [https://github.com/thom311/libnl/commit/c20968a4d6b080c4c03a77dc13d932d8220e6a70]
Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 lib/route/link/vxlan.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/lib/route/link/vxlan.c b/lib/route/link/vxlan.c
index 0603bf590d3f..d0ebc6a81422 100644
--- a/lib/route/link/vxlan.c
+++ b/lib/route/link/vxlan.c
@@ -572,7 +572,7 @@ static int vxlan_put_attrs(struct nl_msg *msg, struct rtnl_link *link)
 		NLA_PUT_U8(msg, IFLA_VXLAN_L3MISS, vxi->vxi_l3miss);
 
 	if (vxi->ce_mask & VXLAN_ATTR_PORT)
-		NLA_PUT_U32(msg, IFLA_VXLAN_PORT, vxi->vxi_port);
+		NLA_PUT_U16(msg, IFLA_VXLAN_PORT, vxi->vxi_port);
 
 	if (vxi->ce_mask & VXLAN_ATTR_UDP_CSUM)
 		NLA_PUT_U8(msg, IFLA_VXLAN_UDP_CSUM, vxi->vxi_udp_csum);
@@ -643,13 +643,15 @@ static int vxlan_compare(struct rtnl_link *link_a, struct rtnl_link *link_b,
 						sizeof(a->vxi_group6)) != 0);
 	diff |= _DIFF(VXLAN_ATTR_LOCAL6, memcmp(&a->vxi_local6, &b->vxi_local6,
 						sizeof(a->vxi_local6)) != 0);
-	diff |= _DIFF(VXLAN_ATTR_UDP_CSUM, a->vxi_proxy != b->vxi_proxy);
+	diff |= _DIFF(VXLAN_ATTR_UDP_CSUM, a->vxi_udp_csum != b->vxi_udp_csum);
 	diff |= _DIFF(VXLAN_ATTR_UDP_ZERO_CSUM6_TX,
-		      a->vxi_proxy != b->vxi_proxy);
+		      a->vxi_udp_zero_csum6_tx != b->vxi_udp_zero_csum6_tx);
 	diff |= _DIFF(VXLAN_ATTR_UDP_ZERO_CSUM6_RX,
-		      a->vxi_proxy != b->vxi_proxy);
-	diff |= _DIFF(VXLAN_ATTR_REMCSUM_TX, a->vxi_proxy != b->vxi_proxy);
-	diff |= _DIFF(VXLAN_ATTR_REMCSUM_RX, a->vxi_proxy != b->vxi_proxy);
+		      a->vxi_udp_zero_csum6_rx != b->vxi_udp_zero_csum6_rx);
+	diff |= _DIFF(VXLAN_ATTR_REMCSUM_TX,
+		      a->vxi_remcsum_tx != b->vxi_remcsum_tx);
+	diff |= _DIFF(VXLAN_ATTR_REMCSUM_RX,
+		      a->vxi_remcsum_rx != b->vxi_remcsum_rx);
 	diff |= _DIFF(VXLAN_ATTR_COLLECT_METADATA,
 		      a->vxi_collect_metadata != b->vxi_collect_metadata);
 	diff |= _DIFF(VXLAN_ATTR_LABEL, a->vxi_label != b->vxi_label);
-- 
2.51.0

